sys/include/firmware.h

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright (C) 2018 Kaspar Schleiser <kaspar@schleiser.de>
+ *                    Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_firmware Firmware Updates
+ * @ingroup     sys
+ * @{
+ *
+ * # riotboot minimal firmware update infrastructure
+ *
+ * ## Overview
+ *
+ * riotboot the name of a minimal bootloader application and infrastructure.  It
+ * consists of
+ *
+ * - the application "riotboot" in "dist/riotboot" which serves as
+ *   minimal bootloader
+ *
+ * - the modules "firmware" and "firmware_update" that are used
+ *   in an image to handle verification and flash-writing
+ *
+ * - the module "ota_coap" which serves as transport for the updates
+ *
+ * - a tool in dist/tools/formware which handles key generation
+ *   and firmware image signing.
+ *
+ * - a couple of make targets to glue everything together
+ *
+ * ## Concept
+ *
+ * riotboot expects the flash to be split in three parts: one for the
+ * bootloader itself, and two slots for images.
+ * Of the latter, one slot holds the running (or latest) image, while the other is
+ * used to write a newer image.
+ *
+ * The bootloader will, on reset, verify the checksum of the image slots, then
+ * choose the one with the highest version number that has a valid checksum.  If
+ * all slots have the same version number, the first one will be booted.  If none
+ * of the slots has a valid checksum, no image will be booted and the bootloader
+ * will enter "while(1);".
+ *
+ * Any image that the bootloader can load is supposed to be signed by the firmware
+ * tool, which will take an image, sign it using tweetnacl prefixes the image with
+ * corresponding metadata.
+ *
+ * The metadata contains
+ *
+ * - "RIOT" as magic number
```

Comment:
Sure. You planning a fork? Four letters are advisable, otherwise you'll have to change the data type.

Reply:
We should avoid flexibility here. It's not a good idea to have several magic numbers, since the bootloader will only get one. If you change it, you need to also reflash the bootlaoder, which is in most cases not possible. The idea is to have something CONSTANT to identify that there's indeed a RIOT (nice magic number btw) image at that address, so we can continue the process of booting.